### ISSUE TITLE
Desktop Zoom

### ISSUE THREAD

**Issue by hpk2987** [chjj/compton#188]
Im not totally sure if this is a feature provided by the composite manager but considering  that compiz,kwin,etc.. are able to zoom an area of the screen, i wanted to know if such feature could be implemented in compton.

**Comment by richardgv** [comment]
@Azelphur:

By the way, KMag has a follow mouse mode ("Settings" menu -> "Follow mouse mode"), although it does not work in the way Compiz's does. I found KMag pretty convenient already in presentation, though.

**Comment by dicktyr** [comment]
I also wish for full-screen zoom/pan which I came to appreciate years ago using a hardware implementation in the graphics card

I find primarily useful to toggle the display between 100% and 200% when I need per-pixel accuracy (and when showing something to friends with poor eyes)

**Comment by prologic** [comment]
As a visually impaired person (*legaly blind*) it is **highly** important that a compositing manager implement "full screen zoom" functionality ala Compiz Fusion's ezoom plugin. Desktop apps such as kmag or gnome-mag are **not** sufficient (*in my experience*). I presently use (*and have done so for >~10 years*) used a combination of CRUX + Xorg + Compilz Fusion (*very old version / pre-Cannonical take-over*) and XFCE.

**Comment by ivan-krukov** [comment]
Very much in favour of this. Would you accept a pull request?

**Comment by prologic** [comment]
If someone has the necessary experience and expertise to implement this (*I don't*) the minimum requirements IHMO are:

- Full Desktop Zoom capabilities (not windowed, etc)
- Zoom levels controlled by (for example) SUPER + SCROLL IN/OUT (Trackpad/Scrollwheel)

This is probably the most basic set of requirements. Once you have this nailed myself and other visually impaired persons could seek to use alternative desktop experiences using Compton and hopefully other window managers.

In a world where I am forced to maintain my own [compiz fusion ports](https://crux.nu/portdb/?a=search&q=compiz) that are ancient versions because Canonical/Ubuntu took over the project and ruined it such that whatever Ubuntu's version is is utterly broken.

**Comment by Azelphur** [comment]
@ivan-krukov if you do do this, if possible please make zoom per-output (eg if I zoom in one one monitor, I don't zoom in on others, same behaviour as compiz fusion)

I'm not visually impaired but I always found this compiz feature incredibly useful. Surprised there aren't more people after this feature honestly.

**Comment by prologic** [comment]
I agree with @Azelphur  on this one;. I've always wanted the ability to use desktop zoom on one screen while it's a normal 1x on the other so that when trying to show someone something they don't have to follow my crazy panning and zooming in/out! :)

**Comment by ivan-krukov** [comment]
Those are all great requirements. It seems that implementing this would take a fair amount of work, though.

**Comment by prologic** [comment]
@ivan-krukov You would be my hero haha :)

**Comment by lnalex** [comment]
I need it for presentations. It's really lacking.

**Comment by ivan-krukov** [comment]
I have looked into a compositor-free solution, e.g. `xrandr` - which is rather broken in several ways.
@richardgv You mention that 'such feature could be implemented'. What would be a place to start?

**Comment by dantix** [comment]
It would be nice to have a feature where you can write a rule for a window and apply certain level of zoom to it. I want to make legacy games on hidpi look even more pixelated than they are now :)

Something like
```
scale=2
scale-rule=[ same type of rules as everywhere else in compton ]

**Comment by jellytux** [comment]
+1 (I will never leave osx unless I can zoom like osx)
- How else can I safely operate my machinery which is 3m away from me whilst I am in bed?!

**Comment by techmouse** [comment]
I was taking a very close look at compton today but the deal breaker was no desktop zoom. I have terrible vision so desktop zoom is a 'must'. I really hope compton gets this addition.

**Comment by ryukinix** [comment]
This will be implemented someday? Compton is better xfwm in various ways, but zoom is very very very VERY useful and I miss it.

**Comment by prologic** [comment]
I dunno what that last comment was about; but Full Desktop Zoom as in what OS X and Compiz Fusion's eZoom provide is vital to any vision impaired person using Linux on the Desktop/Laptop.

As a vision impaired person myself; I now find myself in a position where I can no longer sustain and use Linux on the Desktop/Laptop anymore for lack of good accessibility support in terms of providing full desktop zoom. I also have abandoned Linux on the Desktop/Laptop because I could not figure out a way to get the same crisp/sharp font rendering you get on OS X with Retina displays.

I'm afraid that until Linux catches up with Retina-like technology (however this actually works) and we get back maintained and supported full desktop zoom (Cannonical killed Compilz :/) users with vision impairment such as myself can no longer reasonable use Linux as their primary Desktop/Laptop.

**Comment by ViktorNova** [comment]
@prologic 
Just FYI, KDE's Kwin (which can be used on any desktop environment) has the desktop zoom feature from Compiz, and looks pretty good on high DPI displays too, as many of it's UI elements scale. If you match the DPI in the font settings to the DPI of your screen, it looks really good (fonts suddenly become readable and look really good!)

**Comment by techmouse** [comment]
@prologic 
I think abandoning Linux all together is a step too far in the opposite direction. Linux is still very powerful and personally it very much suits all my needs and gives me the power and control I want over my system. It would take a lot for any OS to tear me away from Linux. But that's subjective.

@ViktorNova
It sure does. It looks great too. But Kwin doesn't support zooming in/out with the mouse wheel and using keystrokes just feels unnatural IMO. But it's a fine suggestion and alternative for anybody who can overlook zooming in/out with the keyboard only.

In any case, one of Compton's big selling points was it's vsync support, seeing how it's one of the few window managers that do support it. But because Compton doesn't have desktop zoom, I'll be sticking with XFWM4 and just have to deal with no vsync for the time being. Hopefully either Compton will get desktop zoom or XFWM4 will get vsync.

**Comment by prologic** [comment]
Not having the same ability to zoom in/out with scroll wheel or touchpad
gestures would be a bit of a PITA :) But thanks for the Kwin tip. To be
fair I'm not abandoning LInux; just finding it very hard to use as a
Desktop OS compared to the accessibility that OS X provides out of the box
which for me I would have to spend a lot of effort getting the same
functionality with Compiz (old versions).


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by ryukinix** [comment]
Any news?

**Comment by nrsimha** [comment]
I would love this feature too. Are there any news, or some way how to get full desktop zoom in i3wm?

**Comment by emugel** [comment]
+1. Maybe we should start a prize for feature requests? 💰

**Comment by ohm314** [comment]
I too would be very interested in such a feature. As many have already said, visually impaired linux users are stuck with a semi-working or outdated version of compiz or rather simplistic magnifier apps like kmag. Nothing near the macos accessibility features!

Is there really no one from the compton developer community interested in at least taking a look at it and figuring out what would be needed to implement such a feature?

**Comment by syslino** [comment]
As a web developer I help myself with xzoom, but i'ts a pain to use. Full Desktop Zoom would be greatly appreciated!

**Comment by tobia** [comment]
To anybody who might want to work on this feature: please add a toggle switch to disable anti-aliasing or bilinear filtering or whatever it's called, while you're at it. As a web developer, my main use case for zoom is to inspect pixel-level rendering and the "filtered" zoom that most compositors provide is a nuisance. I'm stuck with `xzoom` for this reason.

By the way, Apple got it right and added this very option in the accessibility pane.

**Comment by dicktyr** [comment]
> disable anti-aliasing

yes thank you

that is of course essential
to zoom the pixel map as it is
without introducing extra distortion (anti-aliasing)

**Comment by rien333** [comment]
It's honestly pretty bad that people that are visually impaired but want to use FoSS/non-bloat stuff can't do this (in a nice and simple, macOS like manner which is general enough of an experience to fit with basically every DE/WM).  Is there maybe something like a patreon/bounty thing we can setup for this? I myself don't really have the skillset or knowledge to implement this atm

**Comment by joelostblom** [comment]
Full desktop zoom via the mouse wheel would be highly useful when teaching!

[This video](https://www.youtube.com/watch?v=uNcgIf_zo2w) shows how the desktop zooming looks in xfwm4.

**Comment by blackoutworm** [comment]
I agree. Until you implement this feature, I am gonna be stuck with Compiz.
I have tried around 20 zoom apps in Linux. And Compiz is the only useful one where you can mousewheel scroll and customize to your liking. The rest of em are pretty much useless.
Some of them allow desktop zooming, but you have to preset zoom sizes. This doesn't work for people who need to zoom in / out depending on text size, day/night etc.

**Comment by Bielecki** [comment]
Still nothing?

**Comment by frnsys** [comment]
fwiw, there is a more active fork here: https://github.com/yshui/compton
